Terraform unknown git repo dependency name

dwc0011 opened the pull request:
Fixes #4491. 

Since the original issue that I set out to address in [4392](https://github.com/dependabot/dependabot-core/issues/4392) requires unique naming conventions due to valid terraform and we do not have regex to address every single git repo url structure, a generic naming convention is needed when urls are not parsed into sources.

This is a patch to address unknown git repos whose Source.from_url cannot be parsed.  These include AWS CodeCommit and other git repos that do not match on the source regex in [common/lib/dependabot/source.rb](https://github.com/dependabot/dependabot-core/blob/main/common/lib/dependabot/source.rb) and are not a GitHub Enterprise Source, or are an Ignored host.  

In order to create a unique git dependency name the code relies on a parsed Source, in the above cases the object is nil.  I have added url source parsing for AWS CodeCommit, for other un-parseable urls nil is detected, which triggers creating a unique name by using the module name, hardcoded strings, and hash of the url.  The naming convention is similar to parsed sources
`module_name::unknown_provider::repo_name/unknown_name(hash_value)::[ref]`
**module_name** = the name of the terraform module for the git source url
**unknown_provider::repo_name/unknown_name** = a hardcoded string (this is the same for each unknown repo)
**(hash_value)** = Digest::SHA1.hexdigest on the source url
**[ref]** = optional ref

dwc0011 commented:
In addition to comments on the solution if there are other tests that should be implemented in other providers / directories please point me in the right direction.

dwc0011 commented:
Been doing some other testing with this and so far dry run success is good.  Any review/feedback will be greatly appreciated.

dwc0011 commented:
I am finished with my testing and waiting for code review.

jurre commented:
Thanks @dwc0011, have added it to our backlog of things to review

dwc0011 commented:
Now that there is a test for dependency name, once it is merged I will need to add the unknow repo name to the test suite.

Nishnha commented:
Hi @dwc0011,

We recently updated our CI to publish docker built images to GHCR, which requires secrets. Unfortunately, [forked repos do not have access to secrets](https://github.blog/2020-08-03-github-actions-improvements-for-fork-and-pull-request-workflows/), so you will need to either re-open this PR on `dependabot/dependabot-core` as a branch to test against our CI or hold off on this until we find a potential solution for forked PRs running our CI suite.

Sorry for the inconvenience.

lorengordon commented:
Hi @Nishnha, I work with @dwc0011 and he is on holiday this week. Would you be able to push the branch to `dependabot/dependabot-core` yourself and open a new pr?

Nishnha commented:
> Hi @Nishnha, I work with @dwc0011 and he is on holiday this week. Would you be able to push the branch to `dependabot/dependabot-core` yourself and open a new pr?

ah right, only maintainers can open a branch 😬 
I'll have to do a code review and some manual testing before I open a branch, but I will work towards that end.

Hopefully we can find a satisfactory solution forked PRs soon.

lorengordon commented:
Heh, yeah, I've run into this problem also. But I couldn't find a reasonable way around it. So I just avoid any use of secrets in the CI for pull requests on public projects. :/

mctofu commented:
CI for fork PRs was fixed in https://github.com/dependabot/dependabot-core/pull/4569

dwc0011 commented:
> > Hi @Nishnha, I work with @dwc0011 and he is on holiday this week. Would you be able to push the branch to `dependabot/dependabot-core` yourself and open a new pr?
> 
> ah right, only maintainers can open a branch 😬 I'll need to do a code review and some manual testing before I open a branch, but I will work towards that end.
> 
> Hopefully we can find a satisfactory solution for forked PRs soon.

I am back, I see you all figured out a solution and all checks have passed.  Let me know if there are any questions or concerns!  Happy New Year!

Nishnha commented on terraform/spec/dependabot/terraform/file_parser_spec.rb:
```diff
@@ -445,9 +445,17 @@
       end
 
+      before do
+        stub_request(:get, "https://unkown-git-repo-example.com/status").to_return(
```
nit
```suggestion
        stub_request(:get, "https://unknown-git-repo-example.com/status").to_return(
```

Nishnha commented on terraform/spec/dependabot/terraform/file_parser_spec.rb:
```diff
@@ -570,6 +578,44 @@
+
+        it "has the right details for the unknown git repo example" do
+          dependency = subject.find do |x|
+            x.name.include? "unknown_repo::unknown_provider::repo_name/unknown_name("
```
```suggestion
            x.name.include? "unknown_repo::git_provider::git_repo("
```

Nishnha commented on terraform/spec/fixtures/projects/git_tags_013/main.tf:
```diff
@@ -29,3 +29,11 @@ module "dns_dup" {
+}
+
+module "unknown_repo" {
+  source           = "git::https://unkown-git-repo-example.com/reponame/test?ref=1.0.0"
```
nit
```suggestion
  source           = "git::https://unknown-git-repo-example.com/reponame/test?ref=1.0.0"
```

Nishnha commented on terraform/spec/dependabot/terraform_spec.rb:
```diff
@@ -51,5 +51,19 @@
       it { is_expected.to eq("distribution_label::terraform-null-label") }
     end
+
+    context "git unknown source with ref" do
+      let(:name) do
+        "module_name::unknown_provider::repo_name/unknown_name(9685A3B07E8D9C45BE0A3D92B02F13978FB311D8)::tags/0.1.0"
+      end
+
+      it { is_expected.to eq("module_name::unknown_name(9685A3B07E8D9C45BE0A3D92B02F13978FB311D8)") }
+    end
+
+    context "git unknown source without ref" do
+      let(:name) { "module_name::unknown_provider::repo_name/unknown_name(9685A3B07E8D9C45BE0A3D92B02F13978FB311D8)" }
+
+      it { is_expected.to eq("module_name::unknown_name(9685A3B07E8D9C45BE0A3D92B02F13978FB311D8)") }
+    end
```
These `display names` are shown in PR titles and might be too long if we include the SHA. 
If we do not know the source maybe we should just use the `module_name`? What do you think?

This would require further modifying the `register_display_name_builder` in https://github.com/dependabot/dependabot-core/blob/main/terraform/lib/dependabot/terraform.rb

dwc0011 replied:
I can definitely strip the hash, I just assumed it would truncate it if to long.  It doesn't provide any other function other than uniqueness and I believe PRs can have duplicate titles, can you confirm that is true @Nishnha

Nishnha replied:
Yep, PRs can have duplicate titles, and Dependabot uses a unique id (not the title) to interact with PRs anyway.

Removing the hash seems like a good move regardless since we should avoid having an arbitrary SHA in the title of a Dependabot PR. Descriptive is better 😃

Nishnha commented on terraform/spec/dependabot/terraform/file_parser_spec.rb:
```diff
@@ -570,6 +578,44 @@
+            file: "main.tf",
+            source: {
+              type: "git",
+              url: "https://unkown-git-repo-example.com/reponame/test",
```
nit
```suggestion
              url: "https://unknown-git-repo-example.com/reponame/test",
```

Nishnha reviewed (changes requested):
Happy new year!

This looks really good overall 👍🏾 

I found a small typo in the word "unknown" and commented with nit.

Another suggestion I have is instead of using `unknown_provider::repo_name/unknown_name` as the provider and repo, we should use `git_provider::git_repo` since it's more descriptive and terse.

dwc0011 commented:
@Nishnha I concur with the recommended changes and will be addressing them shortly

dwc0011 commented on terraform/lib/dependabot/terraform/file_parser.rb:
```diff
@@ -215,10 +216,15 @@ def registry_source_details_from(source_string)
+        elsif git_source
           name + "::" + git_source.provider + "::" + git_source.repo
+        elsif source[:ref]
+          name + "::unknown_provider::repo_name/unknown_name(" \
```
I am changing this to "git_provider::repo_name/git_repo" so it is the same naming format as known git repos.  The register_display_name_builder splits on the "/"  I could do more checks in that function but I want to simplify that.

Nishnha replied:
sounds good

dwc0011 commented on terraform/lib/dependabot/terraform/file_parser.rb:
```diff
@@ -215,10 +216,15 @@ def registry_source_details_from(source_string)
           name + "::" + git_source.provider + "::" + git_source.repo
+        elsif source[:ref]
+          name + "::unknown_provider::repo_name/unknown_name(" \
+          + Digest::SHA1.hexdigest(source[:url]) + ")::" + source[:ref]
```
Just confirming you are validating this will work and not suggesting anything different, correct?

Nishnha replied:
Correct. That comment is just to explain why we're using a SHA1 digest :)

dwc0011 commented:
@Nishnha I believe I addressed all of the changes, I had some git rebase issues and I fixed them the best I can.  Something has changed in my ruby environment so I can't run the tests locally so that is why there were so many CI runs.  I get a new dev laptop tomorrow so didn't want to try and fix it on this machine.

Nishnha commented:
Great work on this @dwc0011 
This will be released with our next version of Dependabot-Core, which should happen within the next few days!

dwc0011 commented:
> Great work on this @dwc0011 This will be released with our next version of Dependabot-Core, which should happen within the next few days!

Thanks @Nishnha  I appreciate the kind words.  Look forward to the release!
